File: interviewbit-hashing-2-sum.py
class Solution:
    # @param A : tuple of integers
    # @param B : integer
    # @return a list of integers
    def twoSum(self, A, B):
        nums = {}
        for i, num in enumerate(A):
            nums[num] = nums.get(num, []) + [i]

        for i, num in enumerate(A):
            target = B - num
            if target in nums and nums[target][0] < i:
                return [nums[target][0] + 1, i + 1]

        return []

File: interviewbit-hashing-2-sum.java
public class Solution {
    public ArrayList<Integer> twoSum(final List<Integer> a, int b) {
        ArrayList<Integer> returnIndices = new ArrayList<Integer>();
        HashMap<Integer, ArrayList<Integer>> nums = new HashMap<Integer, ArrayList<Integer>>();
        for (int i = 0; i < a.size(); i++) {
            int target = b - a.get(i);
            if (nums.containsKey(target)) {
                returnIndices.add(nums.get(target).get(0) + 1);
                returnIndices.add(i + 1);
                return returnIndices;
            }
            if (nums.containsKey(a.get(i))) {
                nums.get(a.get(i)).add(i);
            } else {
                ArrayList<Integer> al = new ArrayList<Integer>();
                al.add(i);
                nums.put(a.get(i), al);
            }
        }
        return returnIndices;
    }
}
